RPG playerstats: global effects struct and effects functions

sumedhsarwate opened the pull request:
We've implemented the global effects struct, global and player new, init and free functions for effects, and an add_effect_to_game function. Closes #596 and #608.

ecoble commented on include/game-state/stats.h:
```diff
@@ -237,4 +298,48 @@ int free_stats(stats_hash_t *stat);
  */
 int free_stats_global(stats_global_hash_t *stat);
 
+/*
```
The naming and comments on the following functions are a bit confusing, at first it seemed like `effect_free` and `delete_all_effects` were duplicates until I looked at the implementation (same with the global version). I think it would be clearer to name `effect_free` something like `delete_effect` or even `delete_single_effect` so its purpose is clear.

ecoble commented on include/game-state/stats.h:
```diff
@@ -237,4 +298,48 @@ int free_stats(stats_hash_t *stat);
+ * Returns:
+ *  SUCCESS on success, FAILURE if an error occurs.
+ */
+int global_effect_free(effects_global_t *effect);
```
See above comment about naming

ecoble commented on src/game-state/src/stats.c:
```diff
@@ -88,4 +139,62 @@ int free_stats(stats_hash_t *s)
+}
+
+/* See stats.h */
+int effect_free(stat_effect_t *effect)
```
You should probably use hash delete to also remove this effect from the hashtable

ecoble commented on src/game-state/src/stats.c:
```diff
@@ -88,4 +139,62 @@ int free_stats(stats_hash_t *s)
+}
+
+/* See stats.h */
+int global_effect_free(effects_global_t *effect)
```
Same here about deletion

ecoble commented on tests/game-state/test_stats.c:
```diff
@@ -70,4 +70,53 @@ Test(stats, display_stat)
+/* Checks that global_effect_init correctly initializes a global effect struct */
+Test (stats, effect_global_init)
+{
+    effects_global_t effect;
```
It makes more sense for this to be a pointer

ecoble commented on tests/game-state/test_stats.c:
```diff
@@ -70,4 +70,53 @@ Test(stats, display_stat)
+    effects_global_t *global = global_effect_new("speed");
+    cr_assert_not_null(global, "global_effect_new failed");
+
+    stat_effect_t effect;
```
Same here about pointer

sumedhsarwate commented:
I've renamed `effect_free` to `delete_single_effect` and `global_effect_free` to `delete_single_global_effect`. I've also added a parameter to both of them that takes in a pointer to a hash table so that the function can delete the effect from the hash table.

ecoble commented on include/game-state/stats.h:
```diff
@@ -138,6 +150,52 @@ stats_global_t *stats_global_new(char *name, double max);
+ * Returns:
+ *   - Pointer to allocated player effects struct
+ */
+stat_effect_t *stat_effect_new(effects_global_t *global);
```
In keeping with the rest of your naming conventions, it would be better for this to just be `effect_new`

sumedhsarwate replied:
`effect_new` already exists in game_action, so I had to change the name for this function. Should I rename the other functions instead?

ecoble replied:
Ah I see yes that sounds good

ecoble commented on include/game-state/stats.h:
```diff
@@ -237,4 +295,49 @@ int free_stats(stats_hash_t *stat);
+/*
+ * Frees a single player effects struct
+ *
+ * Parameters: 
```
update header comments

ecoble commented on include/game-state/stats.h:
```diff
@@ -237,4 +295,49 @@ int free_stats(stats_hash_t *stat);
+ * Frees a single global effect struct
+ *
+ * Parameters: 
+ * effect: pointer to an effect struct
```
Update header comments here too

ecoble commented on src/game-state/src/stats.c:
```diff
@@ -88,4 +139,63 @@ int free_stats(stats_hash_t *s)
+    assert(effect != NULL);
+    HASH_DEL(hash, effect);
+
+    stat_mod_t *current, *next;
```
Since this is a linked list, you can use utlist functions to delete this list, see here for reference https://troydhanson.github.io/uthash/utlist.html

sumedhsarwate commented:
Summary: the global effects struct was added to dev along with global effect new, init and free functions, player effect new, init and free functions, and an add_effect_to_game function.

ecoble commented:
Issue Score: ✔️++
Comments:
Great work, especially providing updates!
<!-- {"score": "check-plus-plus"} -->
